Workiva/over_react
# UIP-2410: Basic DDC support

## Ultimate problem:
Running OverReact-based code in the DDC resulted in several runtime errors.

Also, the patterns of declaring abstract prop getters was problematic.

## How it was fixed:
- Fix DDC issues within internal code
    - Major fixes:
        - Detect and patch [DDC emulated function name bug](https://github.com/dart-lang/sdk/issues/27647) in UiProps instances
            - This bug prevented `UiProps` classes from having props named `name`
        - Work around [DDC emulated function bug](https://github.com/dart-lang/sdk/issues/29904) encountered when invoking react-dart factories
            - Bump to react-dart `^3.4.0` to consume new `.build()` API
    - Minor fixes:
        - Disable caching (for now) in `getProps` since Expandos not working on certain JS objects
        - Fix mixing in the same props class twice
        - Work around MapViewMixin and Props/StateMapViewMixin errors in DDC
- Fix DDC issues likely to be experienced by consumers
    - Add opt-in transformer flag (`fixDdcAbstractAccessors`) to work around [DDC abstract accessor issue for props/state](https://github.com/dart-lang/sdk/issues/29914)
        - See more info in transformer README changes
- Add tests for:
    - emulated function name bug workaround
        - Some DDC-specific have been added but are disabled for now, until we set up test to run in the DDC
    - abstract accessor transformer workaround


## Testing suggestions:
- Verify all tests pass
    - in Dartium and in the Dart VM: `ddev test` 
    - in dart2js: `ddev test -p chrome` (__not done in CI, needs to be run manually__)
- Pull into WSD and verify that the docs build as expected in strong mode
    - Use branch for this PR: https://github.com/Workiva/web_skin_dart/pull/763

## Potential areas of regression:
- `UiProps` + fluent interface
- transformer


---

> __FYA:__ @greglittlefield-wf @aaronlademann-wf @jacehensley-wf @clairesarsam-wf @joelleibow-wf
